pkg/sql/app_stats.go

```diff
@@ -290,6 +290,9 @@ func (e *Executor) GetScrubbedStmtStats() []roachpb.CollectedStatementStatistics
 		for q, stats := range a.stmts {
 			scrubbed, ok := scrubStmtStatKey(vt, q.stmt)
 			if ok {
+				if len(scrubbed) >= 10000 {
+					scrubbed = scrubbed[:10000]
```

Comment:
1) So there's one thing we haven't made clear in the code before, which you need to know here. The result of scrubbing is currently still valid SQL that can be parsed. I recall that we use this in some places. If we plainly truncate at a fixed offset, then the query risks not be valid any more. I do not yet know 100% if that's an issue. At the least I think we should add a boolean inside the StatementStatisticsKey struct to indicate truncation has occurred.

2) two queries that were different prior to truncation might become equal after truncation. I think we may want a hash of the string in the StatementStatisticsKey too. @dt what do you think?

Reply:
I had imagined we'd truncate server-side, but I think it is okay to do so here as well (i think we _also_ need server-side either way though, to handle reports from the existing binaries).

@knz Do we need to be able to parse after scrubbing? I knew we needed to be able to parse in order to scrub, but IIRC we're okay after scrubbing is done? I might be forgetting something though.

We could also just append something when we truncate (e.g. ellipsis) to indicate truncation... I'm not sure we'd actually use the bool for anything, and would vaguely rather not add churn to the reporting structs unless we need to.

Reply:
yes an ellipsis would definitely work too. But it would also cause a string to be appended, which in turns would cause a heap allocation and a copy of something we just demonstrated was also 10k long. The bool wouldn't need that alloc + copy.

We wanted to keep the string parsable so that we could run the parser to analyze which syntactic features are being used by the query (which rules are being used to traverse the query). If the query is not parsable we'd need to do this analysis by hand.

Reply:
Well, I don't think we can both truncate and retain parseable -- we could only do truncation sever side, so we'd at least have the opportunity to parse and analyze before truncating to save... but i'm not really sure it matters: these are a vanishingly small % of queries that hit this limit, so I'd even just be fine discarding them altogether if that was the quickest fix.

Reply:
All right @dianasaur323 what @dt says, perhaps if the query is too long instead of `scrubbed = scrubbed[:10000]` do `scrubbed = scrubbed[:100] + "... (truncated)"`, i.e. make the query dramatically shorter altogether because we won't be able to analyze it further.

Reply:
Either way, we still need to add long-query truncation in the reg server (both to be defensive and to gracefully handle existing 1.x binaries).

Reply:
ok I agree. @dt in which repo should that truncation issue be filed?

Reply:
I think https://github.com/cockroachlabs/registration/issues/83 is tracking this